furnace.c

```diff
@@ -164,14 +186,133 @@ set_stirrer(bool opt)
+      }
+  }
+#elif CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT || CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+  else if (sscanf(buffer, "spin_dshot %u", &arg) == 1) {
+      const int ret = set_dshot_safe(ctx, arg);
+      if (ret) {
+        char msg[BUF_SIZE];
+        const size_t msg_len =
+          snprintf(msg,
+                  sizeof(msg),
+                  "dshot argument out of scope, should be [%d; %d]!\r\n",
+                  SPIN_COATER_MIN_THROTTLE_COMMAND,
+                  SPIN_COATER_MAX_THROTTLE_COMMAND);
+        feedback(msg, msg_len);
+        DEBUG_printf("Chaning dshot value failed\n");
+        return;
+      }
+
+      if (arg == SPIN_COATER_MIN_THROTTLE_COMMAND) {
+        ctx->spin_coater.spin_state = SPIN_IDLE;
+        cancel_alarm(ctx->spin_coater.spin_timer);
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+        cancel_repeating_timer(&ctx->spin_coater.telemetry_collector_timer);
+        reset_dshot_buffers(&ctx->spin_coater);
+#endif
+        ctx->spin_coater.current_rpm = 0;
+        const char msg[] = "spin coater stopped\r\n";
+        feedback(msg, sizeof(msg)-1);
+      } else {
+        if(ctx->spin_coater.spin_state == SPIN_IDLE)
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+          add_repeating_timer_ms(1,
+                                repeating_dshot_collect_data_callback,
+                                &ctx->spin_coater,
+                                &ctx->spin_coater.telemetry_collector_timer);
+#endif
+
+        ctx->spin_coater.spin_state = SPIN_STARTED_WITH_FORCE_VALUE;
+      }
+  }
+#endif // CONFIG_SPIN_CATER_DSHOT/PWM
+  else if (sscanf(buffer, "spin_start %u %u", &arg, &arg2) ==
+            2) {
+    if(arg > SPIN_COATER_MAX_RPM_VALUE) {
+      char msg[BUF_SIZE];
+      const size_t msg_len = 
+        snprintf(msg,
+                sizeof(msg),
+                "RPM value to high. Max RPM can be %d\r\n",
+                SPIN_COATER_MAX_RPM_VALUE);
+      feedback(msg, msg_len);
+      return;
+    }
+
+    ctx->spin_coater.set_rpm = arg;
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_PWM
+    ctx->spin_coater.pwm_duty = PWM_HEAVY_LOADED_IDLE_DUTY;
+#elif CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT || CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+    ctx->spin_coater.dshot_throttle_val = DSHOT_HEAVY_LOADED_IDLE_DUTY;
+#endif
+    ctx->spin_coater.spin_state = SPIN_STARTED_WITH_TIMER;
+    ctx->spin_coater.spin_timer =
+      add_alarm_in_ms(arg2*MILLISECONDS_PER_SECOND,
+                      timer_spin_callback,
+                      ctx,
+                      false);
+
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+      add_repeating_timer_ms(1,
+                            repeating_dshot_collect_data_callback,
+                            &ctx->spin_coater,
+                            &ctx->spin_coater.telemetry_collector_timer);
+#endif
```

Comment:
If we have logic this complicated, I think It's better to define a separate function for it. This mega handler can already be hard to read.